### PR TITLE
Features/music

### PR DESCRIPTION
**Pull request by LeviathanST**

### CODE REVIEW

File: src/data.rs [hunk 1]
```diff
@@ -48,12 +48,19 @@ pub async fn build_finder(ctx: &Context) -> Result<Nucleo<(u64, String)>> {
...
     pub meme_finder: Mutex<Nucleo<(u64, String)>>,
+    pub http_client: HttpClient,
+    pub songbird: Arc<songbird::Songbird>,
 }
 
 impl Data {
-    pub async fn new(ctx: &Context) -> Result<Self> {
+    pub async fn new(ctx: &Context, songbird: Arc<songbird::Songbird>) -> Result<Self> {
         let meme_finder = Mutex::new(build_finder(ctx).await?);
+        let http_client = HttpClient::new();
```
Comment (akagiyuu) [inline src/data.rs:58]: separated http_client is abundant, use ctx.http() instead

File: src/command/music/play.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    let user_id = ctx.author().id;
+    let bot_id = ctx.cache().current_user().id;
+    let guild_id = ctx.guild_id().unwrap();
+    let manager = &ctx.data().songbird;
+    let vc_state = ctx.guild().unwrap().cmp_voice_channel(&bot_id, &user_id);
+
+    match vc_state {
+        VoiceChannelStates::Same => {
+            ctx.defer().await?;
+            let meta = play_url(ctx.data().http_client.clone(), manager, guild_id, url).await?;
```
Comment (akagiyuu) [inline src/command/music/play.rs]: use ctx.http() instead of ctx.data().http_client

File: src/utils/play.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+use anyhow::Result;
+use poise::serenity_prelude::GuildId;
+use reqwest::Client;
+use songbird::{
+    input::{AuxMetadata, Compose, Input, LiveInput, YoutubeDl},
+    Songbird,
+};
+
+pub async fn play_url(
+    http_client: Client,
```
Comment (akagiyuu) [inline src/utils/play.rs]: can we use reference here?

**Review by akagiyuu — CHANGES_REQUESTED** [review]

File: src/utils/play.rs (hunk 3 above)
Comment (LeviathanST) [reply src/utils/play.rs]: Yes!

File: src/command/music/play.rs (hunk 2 above)
Comment (LeviathanST) [reply src/command/music/play.rs]: ctx.http() return poise's HTTP struct but we need reqwest client here

File: src/data.rs (hunk 1 above)
Comment (LeviathanST) [reply src/data.rs:58]: We are using Songbird, which need reqwest client to create a request to search youtube

**Review by akagiyuu — APPROVED** [review]

**Pull request merged**